src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -3810,4 +3810,179 @@ BOOST_AUTO_TEST_CASE(TransposedConvolutionalLayerOptionalParameterTest)
+  input = arma::mat(6, 2);
+  input << 12 << 443 << arma::endr
+      << 134 << 45 << arma::endr
+      << 11 << 13 << arma::endr
+      << 14 << 55 << arma::endr
+      << 110 << 4 << arma::endr
+      << 1 << 45 << arma::endr;
+
+  result << -0.629337 << 2.14791 << arma::endr
+      << 0.156797 << -0.416694 << arma::endr
+      << -0.63578 << -0.622893 << arma::endr
+      << -0.637481 << 0.4440386 << arma::endr
+      << 1.894857 << -0.901267 << arma::endr
+      << -0.980402 << 0.180253 << arma::endr;
+
+  module1.Forward(input, output);
+  CheckMatrices(result, output, 1e-3);
+
+  runningMean(0) = 16.1792;
+  runningMean(1) = 6.30667;
+  runningVar(0) = 4278.545;
+  runningVar(1) = 201.78576;
+
+  CheckMatrices(runningMean, module1.TrainingMean(), 1e-3);
+  CheckMatrices(runningVar, module1.TrainingVariance(), 1e-3);
+
+  // Check backward function.
+  module1.Backward(input, output, delta);
+
+  deterministicOutput.clear();
+  module1.Deterministic() = true;
+  module1.Forward(input, deterministicOutput);
+
+  result.clear();
+  result << -0.06389128 << 6.525255983 << arma::endr
+      << 1.80125016 << 0.440614188 << arma::endr
+      << -0.07917933155 << -0.0486032421 << arma::endr
+      << 0.5415883265 << 3.427869269 << arma::endr
+      << 7.299709599 << -0.16238263 << arma::endr
+      << -0.3735739237 << 2.723898308 << arma::endr;
+
+  CheckMatrices(result, deterministicOutput, 1e-3);
+}
+
+/**
+ * Batch Normalization layer numerical gradient test.
+ */
+BOOST_AUTO_TEST_CASE(GradientBatchNormWithMiniBatchesTest)
+{
+  // Add function gradient instantiation.
+  // To make this test robust, check it five times.
+  bool pass = false;
+  for (size_t trial = 0; trial < 5; trial++)
+  {
+    struct GradientFunction
+    {
+      GradientFunction()
+      {
+        input = arma::randn(16, 1024);
+        arma::mat target;
+        target.ones(1, 1024);
+
+        model = new FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization>();
+        model->Predictors() = input;
+        model->Responses() = target;
+        model->Add<IdentityLayer<>>();
+        model->Add<Convolution<>>(1, 2, 3, 3, 1, 1, 0, 0, 4, 4);
+        model->Add<BatchNorm<>>(2);
+        model->Add<Linear<>>(2 * 2 * 2, 2);
+        model->Add<LogSoftMax<>>();
+      }
+
+      ~GradientFunction()
+      {
+        delete model;
+      }
+
+      double Gradient(arma::mat& gradient) const
+      {
+        double error = model->Evaluate(model->Parameters(), 0, 1024, false);
+        model->Gradient(model->Parameters(), 0, gradient, 1024);
+        return error;
+      }
+
+      arma::mat& Parameters() { return model->Parameters(); }
+
+      FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization>* model;
+      arma::mat input, target;
+    } function;
+
+    double gradient = CheckGradient(function);
+    std::cout << gradient << std::endl;
+    if (gradient < 8e-2)
+    {
+      pass = true;
+      break;
+    }
```

Comment:
@kartikdutt18 If there is no issue in implementation, I'm fine with this way. With this threshold `8e-2`, how many did you pass this test case? Did you run this test with the different random seed by any chance? Generally, when I made a test case with a stochastic property, I run the test case about 2~3,000 times with the different seed to see whether all the trials pass or not.

Let me know your thoughts

```suggestion
    if (CheckGradient(function) < 8e-2)
    {
      pass = true;
      break;
    }
  }
```

Reply:
I did run it a couple of times (around 7-8), Let me set the seed to NULL and run it 3000 times ans see how many times it passes. I'll get back to you with the results. Thanks.

Reply:
Yes, with this reference, it recommends about 1,000 passes :) but I just run 3,000 times to make test case robust.

Reply:
In thousand runs it passed the tests 972 times (972 / 1000).So, I am increasing the threshold to 0.1 and running the test again.

Reply:
With threshold set to 0.1, I ran it 1500 times, it passed the test in all 1500 runs.

Reply:
Great, it seems that value is enough.